### DISCUSSION TITLE
Flaky tests: One test reflects to another test case run

### DISCUSSION THREAD

**Discussion by tapika** [xunit/xunit#2370] [General]
Let's take an example. We have over 400 tests in one test application.

Mostly unit tests are written in C# and uses C# baseline code. That code in a turn is using C++ code due to higher performance requirement.

At some point we notice that test 125 fails, we investigate it and find out that something happened before that test, and we need to have better cleanup routines in test 123. Ok, that is more or less manageable - we perform cleanup on some dispose method, that is not a problem.

However - next thing next - we notice that test 122 works ok, but it produces some ram corruption, which in a turn reflects to test 125 - it starts to manifest as a problem in test 125.

In theory running all tests in random maybe one approach to solve this - but if problem persists - developer needs to be able to reproduce problematic case. It's possible for example to print out some random seed to output window, and use that seed to control test execution order on both - build machine and on developers machine. 

But randomness is maybe not so good approach, as it needs to be started slowly and then moved all around all tests.

Now the question is maybe following - could xunit test execution granularity controlled on xunit runner level.
xunit runner could have it's own some sort of smart logic. It decides on it's own what tests it executes, when to shutdown / restart test application fully. (Must be full exit process) - and then somehow detect what test might reflect to what test, and reports something about this.

Can something like this be implemented and how ?

Just wondering if xunit runner could be self-learning by itself, but keeping learned data in some centralized place (for example network drive)

**Comment by bradwilson** [comment]
There is no facility for this today.

The randomness that we use is intended to be stable until code changes. We do not expose a randomness seed at this time, so that stability is hard to predict and control.

I've opened issue #2371 to cover this scenario.

**Reply by tapika** [reply]
That's is ok.

But I think main question still remains. If 400 tests take 20 minutes, and we have a case when 1:st test reflects to test 400 - then in worst case scenario you need 20 minutes to reproduce failure, and maybe 20min * 10-20 rounds to fix it permanently.

It would be good if xunit test runner would have logic on it's own to separate and isolate tests, so testing would use less time, and cross test dependency could be located.

So with worst case scenario xunit runner would saw that test 400 fails - then it would rerun same tests with one test in isolation, which end up with 2 exection rounds - 1-399 test first, test 400 next. If all are passed - xunit would report this as "conditionally passed", and continue to move tests from 1-399 to 400-400 pool until we locate which test is causing test 400 to fail. Since everything takes time, this could happen in multiple test rounds. (until cross test dependency is located).

round 2: 1-199 + 199-400
round 3: 1-99 + 100-400
round 4: 1-49 + 50-400
...
round N: 2-399 + 1,400 - found offending test combination, report a failure.

I understand that maybe it would be simplest just to split tests - 200 (10 min) + 200 (10 min) - but it's not clear how many times such splitting needs to be performed.

**Reply by tapika** [reply]
Could something like this be coded into xunit ?
